---
title: "CoVizu: Real-time visualization of SARS-CoV-2 genomic diversity"
author: Art Poon
date: "19/04/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.width=4, fig.height=4)
```



### Objectives
The objectives of this project are:

1. to develop an open-source toolset to efficiently analyze the entire contents of the GISAID SARS-CoV-2 database (making perpetual updates feasible as new genome data become available), and;

2. to generate data visualizations for the purpose of distinguishing between cases of ongoing community transmission versus the new importation of infection from other countries.

3. to continually update these visualizations with new genomic data to identify potentially actionable features for public health.


### Summary

CoVizu is an open source project to develop a `near real time' SARS-CoV-2 genome analysis and visualization system that highlights potential cases of importation from other countries or ongoing community transmission.

The current mode of visualization employed by CoVizu that we are tentatively referring to as a "beadplot":

```{r echo=FALSE, fig.width=7, fig.height=6}
source("../draw-mst.R")
# 5, 14, 15, 17
plot.mst(files[14], mar=c(2,6,0,0), xlab='', threshold=5)
```

#### How to read a beadplot:

* Each horizontal line segment represents a unique SARS-CoV-2 genomic sequence variant.  The emergence of a single new mutation in an infection is sufficient to establish a new variant.  A given variant may be observed multiple times as identical genome sequences, where `identity' is loosely defined to accommodate genomes with incomplete coverage and ambiguous base calls.  (Following GISAID's definition of a "complete genome", we require a minimum sequence length of 29,000 nt.)

* Each circle represents one or more cases of a variant that were sampled on a given date.  The size (area) of the circle is proportional to the number of sequences.

* Cases are arranged in chronological order from left to right.

* Circles are coloured red if *any* of the genomes was sampled in Canada on that date.

  > The first red circle in a series of cases on the same horizontal line represents the first sample of that variant in Canada and implies a new importation.  A series of red circles on the same line implies community transmission.

* Variants are labelled with the name of the earliest genome record.  To reduce clutter, however, it may become necessary to filter labels to only variants that have a minimum number of cases.

* Vertical lines connect variants that are related by a minimum spanning tree, which gives a *rough* approximation of transmission links.  The variant at the bottom terminus of the vertical line is the putative source.  

* It is not feasible to reconstruct accurate links using only genomic data.  However, our objective is to identify population-level events like importations into Canada, not to attribute a transmission to a specific source individual.

* Circles that are the first "bead" on the horizontal line are related to the same ancestral variant if they intersect the same vertical line.  This scenario implies that multiple lineages descend from the same ancestor.

  > If none of the ancestral cases were sampled in Canada, then a first "bead" of a new variant that is red implies importation and mutation.

* The relative location of variants along the vertical axis does not convey any information.  The variants are sorted with respect to the vertical axis such that ancestral variants are always below their "descendant" variants.

#### A cosmopolitan cluster

Let's examine another beadplot from the current dataset:

```{r echo=FALSE, fig.width=5}
setwd('~/git/covizu')
plot.mst(files[15], mar=c(2,7,0,0), xlab='', threshold=0)
```

This cluster "starts" with a genome that was sampled in Ontario (`Canada/ON_PHL8751`); the corresponding point is located in the lower-left of the beadplot.
In total, the genome variant represented by this instance was observed 49 times.
The time period represented by the horizontal line extends past the last date that this variant was observed (March 24) because one of the "descendant" variants (`England/20139042904`) was first observed at a later date (March 28).

This beadplot provides evidence of movement between Ontario and British Columbia (BC).
Specifically, there are two genomes (`Canada/BC_5979789` and `Canada/BC_5275718`) that are putative descendants of the first variant (`Canada/ON_PHL8751`).
Because sampling is very incomplete, we cannot conclude that these lineages moved specifically from Ontario to BC.
The plot also comprises variants that have been sampled in other countries around the world, including the USA, Japan, Saudi Arabia, Finland, Belgium, Australia, Argentina, and the United Kingdom.
Again, incomplete sampling makes it impossible to attribute the source of transmission to Canada or any other country.


#### Caveats

This is an experimental method that is under development.
The vertical lines can mask multiple connections between variants that were mapped to the same day.
Lengths of vertical lines do not correspond to genetic distance.
The absence of any other related genomic variants in the database can result in a connection being drawn between two variants that are separated by a substantial genetic distance, since the method used to deduce epidemiological relationships is forced to find connections to *every* variant in the database.


### Background and rationale


#### Overwhelming data
There is a rapidly accumulating number of genome sequences of severe acute respiratory syndrome coronavirus 2 (SARS-CoV-2) collected at sites around the world, predominantly available through the Global Intiative on Sharing All Influenza Data (GISAID) database.

This figure summarizes the number of SARS-CoV-2 genome records in GISAID since January 1, 2020, up to April 19, 2020, based on the sample collection dates:

```{r, echo=F, fig.width=4, fig.height=3, message=FALSE}
# derived from GISAID acknowledgements table
ack <- read.csv('~/git/covizu/data/acknow.csv', row.names=1)
ack$coldate <- as.Date(ack$coldate)

require(ggfree, quietly=T)

par(xpd=FALSE, mar=c(5,5,1,1))
plot(NA,
     xlim=c(as.Date('2020-01-01'), 
            max(ack$coldate, na.rm=T)),
     ylim=c(1, nrow(ack)),
     xlab='', ylab='Number of genomes',
     xaxt='n', bty='n', col='cadetblue', lwd=2, las=1, cex.axis=0.75)
add.grid(mode='x', bg.col = 'white', fg.col='grey90')
lines(sort(ack$coldate), 1:sum(!is.na(ack$coldate)), 
      type='s', col='cadetblue', lwd=2)


xt <- seq(as.Date('2020-01-01'), max(ack$coldate, na.rm=T), length.out=10)

axis(side=1, at=xt, labels=strftime(xt, '%b %d'), 
     las=2, cex.axis=0.7, mgp=c(3,0.6,0))
title(xlab='Collection date', line=3.)
```

Note that the diminishing number of records in the most recent time period may reflect the delay between sample collection and submission of the genome sequence (including sample processing, next-generation sequencing and analysis).


#### Genomic diversity of SARS-CoV-2

The public release of these genome sequences in near real-time is an unprecedented resource for molecular epidemiology and public health.
For example, [nextstrain](http://nextstrain.org) has been at the forefront of analyzing and communicating the global distribution of SARS-CoV-2 genomic variation.
About four months into the pandemic, a substantial amount of genetic variation has accumulated, as illustrated by the following root-to-tip plot that we generated from a maximum likelihood tree reconstructed (using a double-precision build of FastTree) from an alignment of 2,443 unique genomic variants:

```{r echo=F}
require(ape)
phy <- read.tree('~/git/covizu/data/clusters.ft2.nwk')
info <- read.csv('~/git/covizu/data/clusters.info.csv')

# map sample collection dates to tips of tree
labels <- as.data.frame(matrix(unlist(sapply(unique(info$label), function(x) {
  strsplit(as.character(x), "\\|")[[1]]
  })), ncol=3, byrow=T))
names(labels) <- c('label', 'accession', 'coldate')
labels$coldate <- sapply(labels$coldate, as.Date)
index <- match(phy$tip.label, labels$accession)
phy$coldate <- as.Date(labels$coldate[index], origin='1970-01-01')
phy$label <- labels$label[index]

# reroot the tree on earliest sample from Wuhan 
i <- which.min(phy$coldate[grepl('Wuhan', phy$label)])
root.label <- phy$tip.label[grepl('Wuhan', phy$label)][i]
rooted <- root(phy, root.label)

y <- node.depth.edgelength(rooted)
par(mar=c(5,5,1,1))
plot(rooted$coldate, y[1:Ntip(rooted)]+rnorm(Ntip(phy), 0, 1e-5), bty='n', xaxt='n',
     pch=16, cex=0.5, col='yellowgreen',
     xlab='Collection date', ylab='Root-to-tip distance (ESS)')

xt <- seq(as.Date('2020-01-01'), max(phy$coldate, na.rm=T), length.out=10)
axis(side=1, at=xt, labels=strftime(xt, '%b %d'), 
     las=2, cex.axis=0.7, mgp=c(3,0.6,0))

fit <- lm(y[1:Ntip(rooted)] ~ rooted$coldate)
abline(fit, lwd=2)
```

where the tree was rooted on the earliest genome sampled in Wuhan, China, and the root-to-tip distance was measured in units of the expected number of nucleotide substitutions.
Note that I have added random Gaussian noise ($\sigma=10^{-5}$) to the root-to-tip distances to compensate for the large number of overlapping points.

Based on this tree, we estimate that two genomes sampled at random from around the world would differ at roughly 13.5 positions (ignoring the potential for mutations at the same sites, and comparing pairs of genomes sampled since March 29, 2020):

```{r echo=FALSE, fig.height=3}
# 95% quantile - most recent collection dates
cutoff <- as.Date(quantile(as.integer(phy$coldate), 0.95), origin='1970-01-01')

# approx. genome length 30k
pat.dists <- cophenetic(rooted)
idx <- which(rooted$coldate > cutoff)
temp <- 3e4*pat.dists[idx, idx]
#summary(temp[upper.tri(temp)])

par(mar=c(5,5,1,1))
hist(temp[upper.tri(temp)], main=NA, col='grey', border='white',
     xlab='Expected number of substitutions')
```


#### Time-scaled trees
The central feature of [nextstrain](nextstrain.org) is a reconstruction of a time-scaled phylogeny (a tree-based model of how infections are related by common ancestors back in time).
Geographic locations of samples are mapped onto the tree by colour to communicate the global distribution of the pandemic over time.
However, it is not obvious how useful (actionable) information can be best extracted from these data, particularly in the context of a regional public health authority.

A significant challenge in using a time-scaled tree to visualize the evolutionary (and, to some approximation, epidemiological) relationships among sampled infections is that most of the image is occupied by ancestral lineages whose existence was reconstructed from the observed infections.
We illustrate this with a time-scaled tree that we reconstructed from 2,430 unique genome sequences using the [TreeTime](https://github.com/neherlab/treetime) program released by the nextstrain developers:

```{r fig.height=5, fig.width=6, echo=FALSE, warning=FALSE}
phy <- read.tree('timetree.nwk')  # 2020-04-14
phy <- ladderize(phy, right=F)
require(ggfree)
L <- tree.layout(phy)
plot(L, label='n', mar=c(0,0,0,0))
index <- match(phy$tip.label, labels$accession)
z <- grepl('Canada', labels$label[index])
tip.data <- L$nodes[L$nodes$n.tips==0, ]
points(tip.data$x[!z], tip.data$y[!z], pch=16, cex=0.8)
points(tip.data$x[z], tip.data$y[z], pch=16, col='red2')
```

From this image, we can observe that genomes sampled in Canada are distributed throughout the tree, indicating multiple importations of SARS-CoV-2 into Canada.
However, it is difficult to distinguish between importations and community transmission from the location of Canadian samples in this tree.
This problem is not ameliorated by colouring the branches of the tree to indicate the different countries where samples were collected.

Another limitation of the tree visualization is that it does not convey information about observing the same genome sequence from multiple samples over time.
There is no means to differentiate identical sequences in a phylogeny because there are no phylogenetically informative sites that separate them.
We could extend the tips of the tree to span the time period of sample collection and mark sampled genomes as in the beadplot above.
However, the time scale of sampling identical genomes is relatively short compared to the evolutionary history of the virus that is represented by the tree.


### Workflow

#### Alignment

Generating a multiple sequence alignment (MSA) from thousands of genomes that are about 30,000 nt in length is challenging.
The time complexity of standard heuristic algorithms for alignment is roughly on the order of $n^2L$, where $L$ is the sequence length and $n$ is the number of sequences.
In other words, the time it takes to reconstruct the entire MSA increases quadratically with a growing number of sequences.

We use a pairwise alignment approach to greatly reduce the amount of time required to align the sequences.
By aligning each sequence to a reference genome, and *most importantly* by discarding any sequence insertions relative to the reference genome, we are forcing each sequence into the same coordinate system.
(We refer to this strategy as Procrustean sequence alignment after the myth of Procrustes, the innkeeper who amputates travellers so they will fit into his only bed.)

A major advantage of this approach is that once a query sequence is aligned, it is no longer necessary to re-align it with the addition of other sequences.
In contrast, a conventional MSA must re-process all sequences in the database to accommodate any novel insertions or deletions.

We use an implementation of Altschul and Erikkson's algorithm for pairwise alignment in C that is released as a Python module at:
http://github.com/ArtPoon/gotoh2.


#### Genetic distance analysis

To quickly analyze the evolutionary relationships among genomes in the database, we calculate the pairwise matrix using the Tamura-Nei (TN93) genetic distance
A genetic distance is a formula that maps two sequences to the non-negative real number line, $d(s,t)\rightarrow\mathcal{R}^{+}$, such that $d(s,s)=0$ and $d(s,t) = d(t,s)$.
Simply put, a genetic distance estimates the amount of evolutionary time separating two sequences based on the number of observed differences between them.
The TN93 distance is the most biologically realistic distance that has a closed form expression for rapid calculation.
We use a fast implementation of the TN93 distance in C that is released at http://github.com/veg/tn93.

From this distance matrix, we construct a weighted complete graph in Python using the Python module [networkx](https://networkx.github.io/), where each node represents a genome sequence.
We group nodes that are separated by a distance below $10^{-6}$, which indicates that the genomes are essentially identical.
(This ignores ambiguous nucleotides resulting from incomplete sequencing or low quality base calls.)
Each group is replaced by a single representative genome record with the earliest sample collection date.


#### Minimum spanning tree analysis

The reduced graph that results from merging identical genomes is analyzed with a minimum spanning tree (MST) algorithm, which attempts to find the shortest tree that connects every node in the graph.
By rooting the resulting tree on the genome with the earliest sample collection date, we derive a rough approximation of the epidemiological history of the pandemic.
This approach assumes that the "ancestral nodes" that represent transmission sources have all been sampled in the database.
Thus, an MST-based approach is inherently inferior to a phylogenetic analysis that is able to reconstruct ancestral nodes as latent (unobserved) variables.

On the other hand, the relatively slow rate of molecular evolution of SARS-COV-2, in comparison to the time scale of transmission, supports the assumption that ancestral variants are often sampled.
In addition, using a graph that comprises exclusively of observed variants, instead of having to extrapolate back to unobserved ancestors in a phylogeny, enables us to make more efficient use of space in a visualization.


#### Cluster analysis

Whether we use a phylogeny or an MST, the resulting graph is too large for the purpose of visualization because of the large number of genomic variants that exist in the database.
In other words, we have to partition the database into smaller subsets that are more feasible to display individually.
A phylogeny can be partitioned into subtrees.
However, we are required to decide on quantitative criteria for partitioning the tree that resolves the trade-off between data reduction (the benefit of drawing smaller trees) and loss of information.
Similarly, an MST can be partitioned into some number of subgraphs, each of which can be transformed into a visualization.
This partition is also the outcome of applying some criteria, for which there are no guidelines for setting.

Currently we are using the out-degree of nodes in the MST as the sole criterion for partitioning the graph.
A node with a high out-degree represents a genomc variant that occupies a "central" (average/consensus) position among a relatively large number of other variants.
This is an area that is still under active investigation, and we may decide on a different approach to clustering nodes in the future.
